nomad/state/state_store.go

```diff
@@ -4885,6 +4900,89 @@ func (s *StateStore) updateJobStabilityImpl(index uint64, namespace, jobID strin
 	return s.upsertJobImpl(index, nil, copy, true, txn)
 }
 
+func (s *StateStore) UpdateJobVersionTag(index uint64, namespace string, req *structs.JobApplyTagRequest) error {
+	jobID := req.JobID
+	jobVersion := req.Version
+	tag := req.Tag
+	name := req.Name
+
+	txn := s.db.WriteTxn(index)
+	defer txn.Abort()
+
+	// if no tag is present, this is a tag removal operation.
+	if tag == nil {
+		if err := s.unsetJobVersionTagImpl(index, namespace, jobID, name, txn); err != nil {
+			return err
+		}
+	} else {
+		if err := s.updateJobVersionTagImpl(index, namespace, jobID, jobVersion, tag, txn); err != nil {
+			return err
+		}
+	}
+
+	return txn.Commit()
+}
+
+func (s *StateStore) updateJobVersionTagImpl(index uint64, namespace, jobID string, jobVersion uint64, tag *structs.JobTaggedVersion, txn *txn) error {
+	ws := memdb.NewWatchSet()
+
+	// Note: could use JobByIDAndVersion to get the specific version we want here,
+	// but then we'd have to make a second lookup to make sure we're not applying a duplicate tag name
+	versions, err := s.JobVersionsByID(ws, namespace, jobID)
```

Comment:
The `JobVersionsByID` method needs to accept a watchset because it's being used by reads that might be blocking reads. But this caller is a write so it's not really relevant here. You can pass in `nil` as the watchset, because its `Add` method is designed to handle that safely.

```suggestion
	// Note: could use JobByIDAndVersion to get the specific version we want here,
	// but then we'd have to make a second lookup to make sure we're not applying a duplicate tag name
	versions, err := s.JobVersionsByID(nil, namespace, jobID)
```